beacon_node/operation_pool/src/bron_kerbosch.rs

```diff
@@ -0,0 +1,230 @@
+                .map(|j| ordering[j])
+                .collect();
+            let r = HashTrieSet::default().insert(vi);
+            let x = (0..i)
+                .filter(|j| neighbourhoods[vi].contains(&ordering[*j]))
+                .map(|j| ordering[j])
+                .collect();
+            bron_kerbosch_aux(r, p, x, &neighbourhoods, &mut publish_clique)
+        }
+    }
+
+    cliques
+}
+
+/// A function to the neighbourhoods for all nodes in the list. The neighbourhood `N(a)` of a
+/// vertex `a` in `vertices` is the set of vertices `b` in `vertices` such that
+/// `is_compatible(&a, &b) == true`. The function assumes that `is_compatible` is symmetric,
+/// and returns a symmetric matrix (`Vec<Vec<usize>>`) of indices, where each index corresponds
+/// to the relative vertex in `vertices`.
+fn compute_neigbourhoods<T, F: Fn(&T, &T) -> bool>(
+    vertices: &Vec<T>,
+    is_compatible: F,
+) -> Vec<Vec<usize>> {
+    let mut neighbourhoods = vec![];
+    neighbourhoods.resize_with(vertices.len(), Vec::new);
+    for i in 0..vertices.len() - 1 {
+        for j in i + 1..vertices.len() {
+            if is_compatible(&vertices[i], &vertices[j]) {
+                neighbourhoods[i].push(j);
+                neighbourhoods[j].push(i);
+            }
+        }
+    }
+    neighbourhoods
+}
+
+/// Produces a degeneracy ordering of a set of vertices.
+fn degeneracy_order(num_vertices: usize, neighbourhoods: &[Vec<usize>]) -> Vec<usize> {
+    let mut v: Vec<usize> = (0..num_vertices).collect();
+    let mut o = vec![];
+    // move vertices from v to o in order of minimum degree
+    while !v.is_empty() {
+        let m: Option<usize> = (0..v.len()).min_by_key(|i| neighbourhoods[*i].len());
+        if let Some(i) = m {
+            o.push(v[i]);
+            v.remove(i);
+        } else {
+            break;
+        }
+    }
+    o
+}
+
+/// Auxiliary function to be used in the recursive call of the Bron-Kerbosh algorithm.
+/// Parameters
+///  * `r` - a working clique that is being built
+///  * `p` - a set of candidate vertices to be added to r
+///  * `x` - a set of vertices that have been explored and shouldn't be added to r
+///  * `neighbourhoods` - a data structure to hold the neighbourhoods of each vertex
+///  * `publish_clique` - a callback function to call whenever a clique has been produced
+fn bron_kerbosch_aux<F>(
+    r: HashTrieSet<usize>,
+    mut p: HashTrieSet<usize>,
+    mut x: HashTrieSet<usize>,
+    neighbourhoods: &Vec<Vec<usize>>,
+    publish_clique: &mut F,
+) where
+    F: FnMut(HashTrieSet<usize>),
+{
+    if p.is_empty() && x.is_empty() {
+        publish_clique(r);
+        return;
+    }
+
+    let pivot = find_pivot(&p, &x, neighbourhoods);
+    let pivot_neighbours: HashTrieSet<usize> = neighbourhoods[pivot].iter().cloned().collect();
+    
+    let ip = p.iter().cloned().filter(|e| !pivot_neighbours.contains(e)).collect::<HashTrieSet<_>>();
```

Comment:
The persistent data structures will work best if we can construct the updated ones as modifications of the existing ones. Doing `p.iter().cloned()` will make an entirely new data structure that doesn't share any memory with the previous one.

Instead, this could be something like:

```rust
let mut ip = p.clone();
for e in p.iter() {
    if pivot_neighbours.contains(e) {
       ip.remove_mut(e);
    }
}
```

This is probably a pattern we want to repeat, so we could make a standalone function (like `hash_set_filter`) which implements this.